glide-core/redis-rs/redis/src/commands/cluster_scan.rs

```diff
@@ -284,467 +478,406 @@ impl ScanState {
+///
+/// # Type Parameters
+///
+/// * `C`: The connection type that must implement `ConnectionLike`, `Connect`, `Clone`, `Send`, `Sync`, and `'static`.
+///
 pub(crate) async fn cluster_scan<C>(
-    core: C,
+    core: Arc<InnerCore<C>>,
     cluster_scan_args: ClusterScanArgs,
 ) -> RedisResult<(ScanStateRC, Vec<Value>)>
 where
-    C: ClusterInScan,
+    C: ConnectionLike + Connect + Clone + Send + Sync + 'static,
 {
-    let ClusterScanArgs {
-        scan_state_cursor,
-        match_pattern,
-        count,
-        object_type,
-    } = cluster_scan_args;
-    // If scan_state is None, meaning we start a new scan
-    let scan_state = match scan_state_cursor.get_state_from_wrapper() {
+    // Extract the current scan state cursor and the flag for non-covered slots
+    let scan_state_cursor = &cluster_scan_args.scan_state_cursor;
+    let allow_non_covered_slots = cluster_scan_args.allow_non_covered_slots;
+
+    // Determine the current scan state:
+    // - If an existing scan state is present, use it.
+    // - Otherwise, initiate a new scan.
+    let scan_state = match scan_state_cursor.state_from_wrapper() {
         Some(state) => state,
-        None => match ScanState::initiate_scan(&core).await {
+        None => match ScanState::initiate_scan(&core, allow_non_covered_slots).await {
             Ok(state) => state,
             Err(err) => {
+                // Early return if initiating the scan fails
                 return Err(err);
             }
         },
     };
-    // Send the actual scan command to the address in the scan_state
-    let scan_result = send_scan(
-        &scan_state,
-        &core,
-        match_pattern.clone(),
-        count,
-        object_type.clone(),
-    )
-    .await;
-    let ((new_cursor, new_keys), mut scan_state): ((u64, Vec<Value>), ScanState) = match scan_result
-    {
-        Ok(scan_result) => (from_redis_value(&scan_result)?, scan_state.clone()),
-        Err(err) => match err.kind() {
-            // If the scan command failed to route to the address because the address is not found in the cluster or
-            // the connection to the address cant be reached from different reasons, we will check we want to check if
-            // the problem is problem that we can recover from like failover or scale down or some network issue
-            // that we can retry the scan command to an address that own the next slot we are at.
-            ErrorKind::IoError
-            | ErrorKind::AllConnectionsUnavailable
-            | ErrorKind::ConnectionNotFoundForRoute => {
-                let retry =
-                    retry_scan(&scan_state, &core, match_pattern, count, object_type).await?;
-                (from_redis_value(&retry.0?)?, retry.1)
-            }
-            _ => return Err(err),
-        },
-    };
+    // Send the SCAN command using the current scan state and scan arguments
+    let ((new_cursor, new_keys), mut scan_state): ((u64, Vec<Value>), ScanState) =
```

Comment:
I think that the type can be deduced from the method `try_scan` so there is not need for this part: `: ((u64, Vec<Value>), ScanState)` (did not try to compile it, but I think it will work without it)